### PR TITLE
fix: Fix multi-period DASH with descriptive audio

### LINKED ISSUE #4500: Descriptive audio is not selectable when the adaptation has a role of main, and ABR is enabled
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
3.x, 4.x

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `main`?**
Yes

**Are you using the demo app or your own custom app?**
Both

**If custom app, can you reproduce the issue using our demo app?**
Yes

**What browser and OS are you using?**
macOS 12.6, Chrome 105.0.5195.125

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

**What are the manifest and license server URIs?**

Sent separately

**What configuration are you using?  What is the output of `player.getConfiguration()`?**

Default configuration

**What did you do?**

Load a stream with two audio tracks, both English (`en-US`), one with a role of `main` and one with a role of `description`. When playback begins switch to the description track using `player.selectAudioLanguage('en-US', 'description')`.

**What did you expect to happen?**
Player switches tracks, descriptive audio can be heard, and video ABR performs as expected.

**What actually happened?**

When ABR is enabled (`player.configure({ abr: { enabled: true } })`), the track is selected, but is then shortly changed as the ABR algorithm selects a different bitrate. The following warning appears in the console, and no descriptive audio can be heard:
```
player.js:3819 Changing tracks while abr manager is enabled will likely result in the selected track being overriden. Consider disabling abr before calling selectVariantTrack().
```
When ABR is disabled (`player.configure({ abr: { enabled: false } })`), the descriptive audio can be heard, but all ABR functionality is lost and the player remains at the last bitrate for the remainder of the stream.

### PR DESCRIPTION
**Pull request by littlespex**
Perform the `primary` check after `role` check to improve period flattening on streams with both `main` and `description` audio tracks.

Resolves #4500

### CODE REVIEW

File: lib/util/periods.js [hunk 1]
```diff
@@ -1180,15 +1180,6 @@ shaka.util.PeriodCombiner = class {
...
 
-    // If the language doesn't match, but the candidate is the "primary"
-    // language, then that should be preferred as a fallback.
-    if (!best.primary && candidate.primary) {
-      return true;
-    }
-    if (best.primary && !candidate.primary) {
-      return false;
-    }
-
```
Comment (littlespex) [inline lib/util/periods.js:1191]: @joeyparrish Removing this completely caused a number of issues, but moving it below the role check solves the issue.

File: test/util/periods_unit.js [hunk 2]
```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
     expect(audio2.originalId).toBe('stream2,stream4');
   });
 
+  it('Matches streams with primary in common', async () => {
```
Comment (littlespex) [inline test/util/periods_unit.js:1152]: This test simulates the stream that originally manifested the issue. It is a multi period steam where all periods have the same number of video and audio adaptation sets, and the audio adaptations both have roles, one `main` and one `description`.

File: test/util/periods_unit.js [hunk 3]
```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
...
+    const audio3 = variants[2].audio;
+    expect(audio3.roles).toEqual(['description']);
+    expect(audio3.originalId).toBe('stream2,stream2');
+
+    const audio4 = variants[3].audio;
+    expect(audio4.roles).toEqual(['description']);
+    expect(audio4.originalId).toBe('stream2,stream2');
+  });
+
+  it('Matches streams with mismatched primary', async () => {
```
Comment (littlespex) [inline test/util/periods_unit.js:1215]: This test simulates a similar stream to the one above, but this time with periods representing pre-roll and mid-roll ads that only have `main` audio tracks.

Comment (littlespex) [inline test/util/periods_unit.js:1215]:
```suggestion
  it('Matches streams with mismatched roles', async () => {
```

File: test/util/periods_unit.js (hunk 2 above)
Comment (littlespex) [inline test/util/periods_unit.js:1152]:
```suggestion
  it('Matches streams with roles in common', async () => {
```

File: test/util/periods_unit.js [hunk 4]
```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
     expect(audio2.originalId).toBe('stream2,stream4');
   });
 
+  it('Matches streams with roles in common', async () => {
```
Comment (littlespex) [inline test/util/periods_unit.js:1152]: This test simulates the stream that originally manifested the issue. It is a multi-period steam where all periods have the same number of video and audio adaptation sets, and the audio adaptations both have roles, one `main` and one `description`.

File: test/util/periods_unit.js [hunk 5]
```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
...
+    const audio3 = variants[2].audio;
+    expect(audio3.roles).toEqual(['description']);
+    expect(audio3.originalId).toBe('stream2,stream2');
+
+    const audio4 = variants[3].audio;
+    expect(audio4.roles).toEqual(['description']);
+    expect(audio4.originalId).toBe('stream2,stream2');
+  });
+
+  it('Matches streams with mismatched roles', async () => {
```
Comment (littlespex) [inline test/util/periods_unit.js:1215]: This test simulates a similar stream to the one above, but this time with periods representing pre-roll and mid-roll ads that only have `main` audio tracks.

File: test/util/periods_unit.js [hunk 6]
```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
...
+
+    console.log(variants);
+
+    expect(variants.length).toBe(4);
+
+    expect(variants).toEqual(jasmine.arrayWithExactContents([
+      makeAVVariant(720, 'en'),
+      makeAVVariant(1080, 'en'),
+      makeAVVariant(720, 'en'),
+      makeAVVariant(1080, 'en'),
```
Comment (joeyparrish) [inline test/util/periods_unit.js:1281]: Could you add an optional roles argument to makeAVVariant?  I think this part of the test would be clearer if we could see the expected roles here.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

**Comment by joeyparrish** [comment]
This looks great overall.  I'm excited to include it in the bugfix releases I've been preparing.

File: test/util/periods_unit.js (hunk 6 above)
Comment (littlespex) [reply test/util/periods_unit.js]: Done.

**Review by joeyparrish — APPROVED** [review]

**Review by avelad — APPROVED** [review]

**Pull request merged**